workflow-controller/controller-service/src/main/java/io/hyscale/controller/invoker/ImageBuildComponentInvoker.java

```diff
@@ -113,40 +128,53 @@ protected void doExecute(WorkflowContext context) throws HyscaleException {
         }
     }
 
-    private String getPullImageRegistry(WorkflowContext context) {
-        ServiceSpec serviceSpec = context.getServiceSpec();
-        String stackImage = null;
-        try {
-            BuildSpec buildSpec = serviceSpec.get(
-                    HyscaleSpecFields.getPath(HyscaleSpecFields.image, HyscaleSpecFields.buildSpec), BuildSpec.class);
-            if (buildSpec != null) {
-                stackImage = buildSpec.getStackImage();
-            }
-        } catch (HyscaleException e) {
-            logger.error("Error while getting build spec for stack image");
-        }
-        return stackImage != null ? stackImage.split("/")[0] : stackImage;
+    private List<String> getImageRegistries(ServiceSpec serviceSpec) throws HyscaleException {
+        List<String> registriesList = new ArrayList<>();
+        String pushRegistry = serviceSpec
+                .get(HyscaleSpecFields.getPath(HyscaleSpecFields.image, HyscaleSpecFields.registry), String.class);
+        registriesList.add(pushRegistry);
+        registriesList.addAll(getPullRegistries(serviceSpec));
+        return registriesList;
     }
 
-    private boolean isStackImage(WorkflowContext context) {
-        Boolean stackAsServiceImage = (Boolean) context.getAttribute(WorkflowConstants.STACK_AS_SERVICE_IMAGE);
-        return stackAsServiceImage == null ? false : stackAsServiceImage;
+    /**
+     * Pull registries include:
+     * Stack image from build spec or
+     * From field in user docker file
+     * @param serviceSpec
+     * @return
+     */
+    private List<String> getPullRegistries(ServiceSpec serviceSpec) {
+        List<String> pullRegistries = new ArrayList<>();
+        String stackImage = stackImageProvider.getStackImageFromBuildSpec(serviceSpec);
+        if (stackImage != null) {
+            // buildSpec
+            pullRegistries.add(stackImage.split("/")[0]);
+        } else {
+            // Dockerfile
+            List<String> stackImages = stackImageProvider.getStackImagesFromDockerfile(serviceSpec);
+            if (CollectionUtils.isNotEmpty(stackImages)) {
+                Set<String> stackRegistries = stackImages.stream().filter(registry -> registry.split("/").length > 1)
+                .map(registry -> registry.split("/")[0]).collect(Collectors.toSet());
+                pullRegistries.addAll(stackRegistries);
+            }
+        }
+        return pullRegistries;
     }
 
     @Override
     protected void onError(WorkflowContext context, HyscaleException he) throws HyscaleException {
         WorkflowLogger.header(ControllerActivity.ERROR);
-        WorkflowLogger.error(ControllerActivity.CAUSE, he != null ? he.getMessage() : ImageBuilderErrorCodes.FAILED_TO_BUILD_AND_PUSH_IMAGE.getMessage());
-        context.addAttribute(WorkflowConstants.ERROR_MESSAGE, (he != null) ? he.getMessage() : ImageBuilderErrorCodes.FAILED_TO_BUILD_AND_PUSH_IMAGE.getMessage());
+        WorkflowLogger.error(ControllerActivity.CAUSE,
```

Comment:
In case of structured o/p?

Reply:
Haven't made any changes for that scenario, so it is being handled as it was earlier.